// telnyx/dist/resources/Faxes.js
import TelnyxResource from '../TelnyxResource.js';
import * as utils from '../utils.js';
const telnyxMethod = TelnyxResource.method;
function transformResponseData(response, telnyx) {
    return utils.addResourceToResponseData(response, telnyx, 'faxes', {
        del: telnyxMethod({
            method: 'DELETE',
            path: '/{faxId}',
            urlParams: ['faxId'],
            paramsValues: [response.data.id],
            paramsNames: ['id'],
        }),
    });
}
export const Faxes = TelnyxResource.extend({
    path: 'faxes',
    list: telnyxMethod({
        method: 'GET',
        methodType: 'list',
    }),
    create: telnyxMethod({
        method: 'POST',
        transformResponseData: transformResponseData,
    }),
    send: telnyxMethod({
        method: 'POST',
        transformResponseData: transformResponseData,
    }),
    retrieve: telnyxMethod({
        method: 'GET',
        path: '/{id}',
        urlParams: ['id'],
        transformResponseData: transformResponseData,
    }),
});
