// telnyx/dist/resources/FqdnConnections.js
import TelnyxResource from '../TelnyxResource.js';
import * as utils from '../utils.js';
const telnyxMethod = TelnyxResource.method;
function transformResponseData(response, telnyx) {
    return utils.addResourceToResponseData(response, telnyx, 'fqdnConnections', {
        del: telnyxMethod({
            method: 'DELETE',
            path: '/{fqdnConnectionId}',
            urlParams: ['fqdnConnectionId'],
            paramsValues: [response.data.id],
            paramsNames: ['id'],
        }),
        update: telnyxMethod({
            method: 'PATCH',
            path: '/{fqdnConnectionId}',
            urlParams: ['fqdnConnectionId'],
            paramsValues: [response.data.id],
            paramsNames: ['id'],
        }),
    });
}
export const FqdnConnections = TelnyxResource.extend({
    path: 'fqdn_connections',
    list: telnyxMethod({
        method: 'GET',
        methodType: 'list',
    }),
    create: telnyxMethod({
        method: 'POST',
        transformResponseData: transformResponseData,
    }),
    retrieve: telnyxMethod({
        method: 'GET',
        path: '/{id}',
        urlParams: ['id'],
        transformResponseData: transformResponseData,
    }),
});
