// telnyx/dist/resources/MessagingProfiles.js
import TelnyxResource from '../TelnyxResource.js';
import * as utils from '../utils.js';
const telnyxMethod = TelnyxResource.method;
const MESSAGING_PROFILES_COMMANDS = ['phone_numbers', 'short_codes'];
function getSpec(messagingProfileId) {
    return function (methodName) {
        return {
            method: 'GET',
            path: `/{messagingProfileId}/${methodName}`,
            urlParams: ['messagingProfileId'],
            paramsValues: [messagingProfileId],
            paramsNames: ['messagingProfileId'],
            methodType: 'list',
        };
    };
}
const transformResponseData = (response, telnyx) => {
    const methods = utils.createNestedMethods(telnyxMethod, MESSAGING_PROFILES_COMMANDS, getSpec(response.data.id));
    methods.del = telnyxMethod({
        method: 'DELETE',
        path: '/{messagingProfileId}',
        urlParams: ['messagingProfileId'],
        paramsValues: [response.data.id],
        paramsNames: ['messagingProfileId'],
    });
    methods.listAutorespConfigs = telnyxMethod({
        method: 'GET',
        path: '/{profileId}/autoresp_configs',
        urlParams: ['profileId'],
        paramsValues: [response.data.id],
        paramsNames: ['profileId'],
        methodType: 'list',
    });
    methods.createAutorespConfig = telnyxMethod({
        method: 'POST',
        path: '/{profileId}/autoresp_configs',
        urlParams: ['profileId'],
        paramsValues: [response.data.id],
        paramsNames: ['profileId'],
        methodType: 'create',
    });
    methods.delAutorespConfig = telnyxMethod({
        method: 'DELETE',
        path: '/{profileId}/autoresp_configs/{autorespCfgId}',
        paramsValues: [response.data.id],
        urlParams: ['profileId', 'autorespCfgId'],
        paramsNames: ['profileId', 'autorespCfgId'],
    });
    methods.retrieveAutorespConfig = telnyxMethod({
        method: 'GET',
        path: '/{profileId}/autoresp_configs/{autorespCfgId}',
        paramsValues: [response.data.id],
        urlParams: ['profileId', 'autorespCfgId'],
        paramsNames: ['profileId', 'autorespCfgId'],
        methodType: 'retrieve',
    });
    methods.updateAutorespConfig = telnyxMethod({
        method: 'PUT',
        path: '/{profileId}/autoresp_configs/{autorespCfgId}',
        paramsValues: [response.data.id],
        urlParams: ['profileId', 'autorespCfgId'],
        paramsNames: ['profileId', 'autorespCfgId'],
    });
    methods.retrieveMetrics = telnyxMethod({
        method: 'GET',
        path: '/{messagingProfileId}/metrics',
        urlParams: ['messagingProfileId'],
        paramsValues: [response.data.id],
        paramsNames: ['messagingProfileId'],
        methodType: 'retrieve',
    });
    return utils.addResourceToResponseData(response, telnyx, 'messagingProfiles', methods);
};
export const MessagingProfiles = TelnyxResource.extend({
    path: 'messaging_profiles',
    includeBasic: ['list', 'del', 'update'],
    create: telnyxMethod({
        method: 'POST',
        transformResponseData: transformResponseData,
    }),
    retrieve: telnyxMethod({
        method: 'GET',
        path: '/{id}',
        urlParams: ['id'],
        transformResponseData: transformResponseData,
    }),
    listPhoneNumbers: telnyxMethod({
        method: 'GET',
        path: '/{messagingProfileId}/phone_numbers',
        urlParams: ['messagingProfileId'],
        methodType: 'list',
    }),
    listShortCodes: telnyxMethod({
        method: 'GET',
        path: '/{messagingProfileId}/short_codes',
        urlParams: ['messagingProfileId'],
        methodType: 'list',
    }),
    listAutorespConfigs: telnyxMethod({
        method: 'GET',
        path: '/{profileId}/autoresp_configs',
        urlParams: ['profileId'],
        paramsNames: ['profileId'],
        methodType: 'list',
    }),
    createAutorespConfig: telnyxMethod({
        method: 'POST',
        path: '/{profileId}/autoresp_configs',
        urlParams: ['profileId'],
        paramsNames: ['profileId'],
        methodType: 'create',
    }),
    delAutorespConfig: telnyxMethod({
        method: 'DELETE',
        path: '/{profileId}/autoresp_configs/{autorespCfgId}',
        urlParams: ['profileId', 'autorespCfgId'],
        paramsNames: ['profileId', 'autorespCfgId'],
    }),
    retrieveAutorespConfig: telnyxMethod({
        method: 'GET',
        path: '/{profileId}/autoresp_configs/{autorespCfgId}',
        urlParams: ['profileId', 'autorespCfgId'],
        paramsNames: ['profileId', 'autorespCfgId'],
        methodType: 'retrieve',
    }),
    updateAutorespConfig: telnyxMethod({
        method: 'PUT',
        path: '/{profileId}/autoresp_configs/{autorespCfgId}',
        urlParams: ['profileId', 'autorespCfgId'],
        paramsNames: ['profileId', 'autorespCfgId'],
    }),
});
