// telnyx/dist/resources/TexmlApplications.js
import TelnyxResource from '../TelnyxResource.js';
import * as utils from '../utils.js';
const telnyxMethod = TelnyxResource.method;
function transformResponseData(response, telnyx) {
    return utils.addResourceToResponseData(response, telnyx, 'texmlApplications', {
        del: telnyxMethod({
            method: 'DELETE',
            path: '/{texmlApplicationId}',
            urlParams: ['texmlApplicationId'],
            paramsValues: [response.data.id],
            paramsNames: ['texmlApplicationId'],
        }),
        update: telnyxMethod({
            method: 'PATCH',
            path: '/{texmlApplicationId}',
            urlParams: ['texmlApplicationId'],
            paramsValues: [response.data.id],
            paramsNames: ['texmlApplicationId'],
        }),
    });
}
export const TexmlApplications = TelnyxResource.extend({
    path: 'texml_applications',
    includeBasic: ['list', 'update', 'del'],
    create: telnyxMethod({
        method: 'POST',
        transformResponseData: transformResponseData,
    }),
    retrieve: telnyxMethod({
        method: 'GET',
        path: '/{id}',
        urlParams: ['id'],
        transformResponseData: transformResponseData,
    }),
});
