// telnyx/dist/resources/Campaign.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Campaign = void 0;
const TelnyxResource_js_1 = __importDefault(require("../TelnyxResource.js"));
const utils = __importStar(require("../utils.js"));
const telnyxMethod = TelnyxResource_js_1.default.method;
function transformResponseData(response, telnyx) {
    return utils.addResourceToResponseData(response, telnyx, '/campaign', {
        del: telnyxMethod({
            method: 'DELETE',
            path: '/campaign',
            urlParams: ['campaign'],
            paramsValues: [response.data.id],
            paramsNames: ['id'],
        }),
        update: telnyxMethod({
            method: 'PATCH',
            path: '/campaign',
            urlParams: ['campaign'],
            paramsValues: [response.data.id],
            paramsNames: ['id'],
        }),
    });
}
exports.Campaign = TelnyxResource_js_1.default.extend({
    path: 'campaign',
    includeBasic: ['create', 'list'],
    acceptShareCampaign: telnyxMethod({
        method: 'POST',
        transformResponseData: transformResponseData,
        path: '/acceptSharing/{campaignId}',
    }),
    getCampaignCost: telnyxMethod({
        method: 'GET',
        transformResponseData: transformResponseData,
        path: '/usecase/cost',
    }),
    retrieveCampaignId: telnyxMethod({
        method: 'GET',
        path: '/{campaignId}',
        urlParams: ['campaignId'],
        transformResponseData: transformResponseData,
    }),
    updateCampaignId: telnyxMethod({
        method: 'PATCH',
        path: '/{campaignId}',
        urlParams: ['campaignId'],
        transformResponseData: transformResponseData,
    }),
    deactivateCampaignId: telnyxMethod({
        method: 'DELETE',
        path: '/{campaignId}',
        urlParams: ['campaignId'],
        transformResponseData: transformResponseData,
    }),
    retrieveMnoMetadata: telnyxMethod({
        method: 'GET',
        path: '/{campaignId}/mnoMetadata',
        urlParams: ['campaignId'],
        transformResponseData: transformResponseData,
    }),
    retrieveOperationStatus: telnyxMethod({
        method: 'GET',
        path: '/{campaignId}/operationStatus',
        urlParams: ['campaignId'],
        transformResponseData: transformResponseData,
    }),
    retrieveOsrCampaignAttributes: telnyxMethod({
        method: 'GET',
        path: '/{campaignId}/osr/attributes',
        urlParams: ['campaignId'],
    }),
    retrieveSharingStatus: telnyxMethod({
        method: 'GET',
        path: '/{campaignId}/sharing',
        urlParams: ['campaignId'],
    }),
});
