// telnyx/dist/resources/DialogflowConnections.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DialogflowConnections = void 0;
const TelnyxResource_js_1 = __importDefault(require("../TelnyxResource.js"));
const telnyxMethod = TelnyxResource_js_1.default.method;
exports.DialogflowConnections = TelnyxResource_js_1.default.extend({
    path: 'dialogflow_connections',
    retrieve: telnyxMethod({
        method: 'GET',
        path: '/{connection_id}',
        urlParams: ['connection_id'],
        methodType: 'retrieve',
    }),
    create: telnyxMethod({
        method: 'POST',
        path: '/{connection_id}',
        urlParams: ['connection_id'],
    }),
    update: telnyxMethod({
        method: 'PUT',
        path: '/{connection_id}',
        urlParams: ['connection_id'],
    }),
    del: telnyxMethod({
        method: 'DELETE',
        path: '/{connection_id}',
        urlParams: ['connection_id'],
    }),
});
