// telnyx/dist/resources/GlobalIps.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalIps = void 0;
const TelnyxResource_js_1 = __importDefault(require("../TelnyxResource.js"));
const telnyxMethod = TelnyxResource_js_1.default.method;
exports.GlobalIps = TelnyxResource_js_1.default.extend({
    path: 'global_ips',
    includeBasic: ['delete', 'list', 'retrieve'],
    DeleteGlobalIpAssignment: telnyxMethod({
        method: 'DELETE',
        path: '/global_ip_assignments/{id}',
    }),
    DeleteGlobalIpHealthCheck: telnyxMethod({
        method: 'DELETE',
        path: '/global_ip_health/checks/{id}',
    }),
    DeleteGlobalIp: telnyxMethod({
        method: 'DELETE',
        path: '/global/ips/{id}',
    }),
    ListGlobalIpAssignments: telnyxMethod({
        method: 'GET',
        path: '/global_ip_assignments',
    }),
    ListGlobalIpProtocols: telnyxMethod({
        method: 'GET',
        path: '/global_ip_protocols',
    }),
    ListGlobalIps: telnyxMethod({
        method: 'GET',
        path: '/global/ips',
    }),
    ListGlobalIpHealthChecks: telnyxMethod({
        method: 'GET',
        path: '/global_ip_health/checks',
    }),
    ListGlobalIpAllowedPorts: telnyxMethod({
        method: 'GET',
        path: '/global_ip_allowed/ports',
    }),
    ListGlobalIpHealthCheckTypes: telnyxMethod({
        method: 'GET',
        path: '/global_ip_health/check/types',
    }),
});
