// telnyx/dist/resources/PhoneNumberAssignmentByProfile.js
import TelnyxResource from '../TelnyxResource.js';
import * as utils from '../utils.js';
const telnyxMethod = TelnyxResource.method;
function transformResponseData(response, telnyx) {
    return utils.addResourceToResponseData(response, telnyx, 'phoneNumberAssignmentByProfile', {
        del: telnyxMethod({
            method: 'DELETE',
            path: '/{phoneNumberAssignmentByProfileId}',
            urlParams: ['phoneNumberAssignmentByProfileId'],
            paramsValues: [response.data.id],
            paramsNames: ['id'],
        }),
        update: telnyxMethod({
            method: 'PATCH',
            path: '/{phoneNumberAssignmentByProfileId}',
            urlParams: ['phoneNumberAssignmentByProfileId'],
            paramsValues: [response.data.id],
            paramsNames: ['id'],
        }),
    });
}
export const PhoneNumberAssignmentByProfile = TelnyxResource.extend({
    path: 'phoneNumberAssignmentByProfile',
    create: telnyxMethod({
        method: 'POST',
        transformResponseData: transformResponseData,
    }),
    retrieve: telnyxMethod({
        method: 'GET',
        path: '/{taskId}',
        urlParams: ['taskId'],
        transformResponseData: transformResponseData,
    }),
    retrieveTaskStatus: telnyxMethod({
        method: 'GET',
        path: '/{taskId}',
        urlParams: ['taskId'],
        transformResponseData: transformResponseData,
    }),
    phoneNumbers: telnyxMethod({
        method: 'GET',
        path: '/{taskId}/phoneNumbers',
        urlParams: ['taskId'],
        transformResponseData: transformResponseData,
    }),
    retrievePhoneNumberStatus: telnyxMethod({
        method: 'GET',
        path: '/{taskId}/phoneNumbers',
        urlParams: ['taskId'],
        transformResponseData: transformResponseData,
    }),
});
