// telnyx/dist/resources/PhoneNumbers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhoneNumbers = void 0;
const TelnyxResource_js_1 = __importDefault(require("../TelnyxResource.js"));
const utils = __importStar(require("../utils.js"));
const PhoneNumbersMessaging_js_1 = require("./PhoneNumbersMessaging.js");
const PhoneNumbersVoice_js_1 = require("./PhoneNumbersVoice.js");
const telnyxMethod = TelnyxResource_js_1.default.method;
function transformResponseData(response, telnyx) {
    return utils.addResourceToResponseData(response, telnyx, 'phoneNumbers', {
        del: telnyxMethod({
            method: 'DELETE',
            path: '/{intId}',
            urlParams: ['intId'],
            paramsValues: [response.data.id],
            paramsNames: ['intId'],
        }),
        update: telnyxMethod({
            method: 'PATCH',
            path: '/{intId}',
            urlParams: ['intId'],
            paramsValues: [response.data.id],
            paramsNames: ['intId'],
        }),
    });
}
exports.PhoneNumbers = TelnyxResource_js_1.default.extend({
    path: 'phone_numbers',
    includeBasic: ['list', 'update', 'del'],
    nestedResources: {
        Messaging: PhoneNumbersMessaging_js_1.PhoneNumbersMessaging,
        Voice: PhoneNumbersVoice_js_1.PhoneNumbersVoice,
    },
    retrieve: telnyxMethod({
        method: 'GET',
        path: '/{id}',
        urlParams: ['id'],
        transformResponseData: transformResponseData,
    }),
    retrieveVoiceSettings: telnyxMethod({
        method: 'GET',
        path: '/{intId}/voice',
        urlParams: ['intId'],
        paramsNames: ['intId'],
        methodType: 'retrieve',
    }),
    updateVoiceSettings: telnyxMethod({
        method: 'PATCH',
        path: '/{intId}/voice',
        urlParams: ['intId'],
        paramsNames: ['intId'],
    }),
    retrieveMessagingSettings: telnyxMethod({
        method: 'GET',
        path: '/{intId}/messaging',
        urlParams: ['intId'],
        paramsNames: ['intId'],
        methodType: 'retrieve',
    }),
    updateMessagingSettings: telnyxMethod({
        method: 'PATCH',
        path: '/{intId}/messaging',
        urlParams: ['intId'],
        paramsNames: ['intId'],
    }),
    enableEmergencySettings: telnyxMethod({
        method: 'POST',
        path: '/{intId}/actions/enable_emergency',
        urlParams: ['intId'],
        paramsNames: ['intId'],
    }),
    retrieveVoicemail: telnyxMethod({
        method: 'GET',
        path: '/{phone_number_id}/voicemail',
        urlParams: ['phone_number_id'],
        paramsNames: ['phone_number_id'],
        methodType: 'retrieve',
    }),
    createVoicemail: telnyxMethod({
        method: 'POST',
        path: '/{phone_number_id}/voicemail',
        urlParams: ['phone_number_id'],
        paramsNames: ['phone_number_id'],
    }),
    updateVoicemail: telnyxMethod({
        method: 'PATCH',
        path: '/{phone_number_id}/voicemail',
        urlParams: ['phone_number_id'],
        paramsNames: ['phone_number_id'],
    }),
});
