// telnyx/dist/resources/PhoneNumberBlocksBackgroundJobs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhoneNumberBlocksBackgroundJobs = void 0;
const TelnyxResource_js_1 = __importDefault(require("../TelnyxResource.js"));
const telnyxMethod = TelnyxResource_js_1.default.method;
exports.PhoneNumberBlocksBackgroundJobs = TelnyxResource_js_1.default.extend({
    path: 'phone_number_blocks_background_jobs',
    includeBasic: ['create', 'list', 'retrieve'],
    CreatePhoneNumberBlockDeletionJob: telnyxMethod({
        method: 'POST',
        path: 'phone_number_blocks_jobs/delete/phone_number_block',
    }),
    GetPhoneNumberBlocksJob: telnyxMethod({
        method: 'GET',
        path: 'phone_number_blocks_jobs{id}',
        urlParams: ['id'],
    }),
    ListPhoneNumberBlocksJobs: telnyxMethod({
        method: 'GET',
        path: '/phone_number_blocks_jobs',
    }),
});
