// telnyx/dist/resources/PortingOrders.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PortingOrders = void 0;
const TelnyxResource_js_1 = __importDefault(require("../TelnyxResource.js"));
const utils = __importStar(require("../utils.js"));
const telnyxMethod = TelnyxResource_js_1.default.method;
function transformResponseData(response, telnyx) {
    return utils.addResourceToResponseData(response, telnyx, 'portingOrders', {
        del: telnyxMethod({
            method: 'DELETE',
            path: '/{portingOrderId}',
            urlParams: ['portingOrderId'],
            paramsValues: [response.data.id],
            paramsNames: ['portingOrderId'],
        }),
        update: telnyxMethod({
            method: 'PATCH',
            path: '/{portingOrderId}',
            urlParams: ['portingOrderId'],
            paramsValues: [response.data.id],
            paramsNames: ['portingOrderId'],
        }),
        activate: telnyxMethod({
            method: 'POST',
            path: '/{porting_order_id}/actions/activate',
            urlParams: ['porting_order_id'],
            paramsValues: [response.data.id],
            paramsNames: ['porting_order_id'],
            methodType: 'create',
        }),
        cancel: telnyxMethod({
            method: 'POST',
            path: '/{porting_order_id}/actions/activate',
            urlParams: ['porting_order_id'],
            paramsValues: [response.data.id],
            paramsNames: ['porting_order_id'],
            methodType: 'create',
        }),
        confirm: telnyxMethod({
            method: 'POST',
            path: '/{porting_order_id}/actions/activate',
            urlParams: ['porting_order_id'],
            paramsValues: [response.data.id],
            paramsNames: ['porting_order_id'],
            methodType: 'create',
        }),
        share: telnyxMethod({
            method: 'POST',
            path: '/{porting_order_id}/actions/activate',
            urlParams: ['porting_order_id'],
            paramsValues: [response.data.id],
            paramsNames: ['porting_order_id'],
            methodType: 'create',
        }),
    });
}
exports.PortingOrders = TelnyxResource_js_1.default.extend({
    path: 'porting_orders',
    includeBasic: ['list', 'del', 'update'],
    create: telnyxMethod({
        method: 'POST',
        methodType: 'create',
    }),
    // include_phone_numbers query param
    retrieve: telnyxMethod({
        method: 'GET',
        path: '/{id}',
        urlParams: ['id'],
        transformResponseData: transformResponseData,
    }),
    listExceptionTypes: telnyxMethod({
        method: 'GET',
        path: '/exception_types',
        methodType: 'list',
    }),
    listPhoneNumberConfigurations: telnyxMethod({
        method: 'GET',
        path: '/phone_number_configurations',
        methodType: 'list',
        urlParams: ['id'],
    }),
    createPhoneNumberConfigurations: telnyxMethod({
        method: 'POST',
        path: '/phone_number_configurations',
        methodType: 'create',
    }),
    listActivationJobs: telnyxMethod({
        method: 'GET',
        path: '/{id}/activation_jobs',
        methodType: 'list',
        urlParams: ['id'],
    }),
    retrieveActivationJob: telnyxMethod({
        method: 'GET',
        path: '/{id}/activation_jobs/{activationJobId}',
        urlParams: ['id', 'activationJobId'],
        paramsNames: ['id', 'activationJobId'],
    }),
    listAllowedFocWindows: telnyxMethod({
        method: 'GET',
        path: '/{id}/allowed_foc_windows',
        methodType: 'list',
        urlParams: ['id'],
    }),
    listAdditionalDocuments: telnyxMethod({
        method: 'GET',
        path: '/{id}/additional_documents',
        methodType: 'list',
        urlParams: ['id'],
    }),
    createAdditionalDocuments: telnyxMethod({
        method: 'POST',
        path: '/{porting_order_id}/additional_documents',
        urlParams: ['porting_order_id'],
        paramsNames: ['porting_order_id'],
        methodType: 'create',
    }),
    delAdditionalDocuments: telnyxMethod({
        method: 'DELETE',
        path: '/{porting_order_id}/additional_documents/{additional_document_id}',
        urlParams: ['porting_order_id', 'additional_document_id'],
        paramsNames: ['porting_order_id', 'additional_document_id'],
    }),
    retrieveLoaTemplate: telnyxMethod({
        method: 'GET',
        path: '/{id}/loa_template',
        urlParams: ['id'],
        methodType: 'retrieve',
    }),
    activate: telnyxMethod({
        method: 'POST',
        path: '/{porting_order_id}/actions/activate',
        urlParams: ['porting_order_id'],
        paramsNames: ['porting_order_id'],
        methodType: 'create',
    }),
    cancel: telnyxMethod({
        method: 'POST',
        path: '/{porting_order_id}/actions/cancel',
        urlParams: ['porting_order_id'],
        paramsNames: ['porting_order_id'],
        methodType: 'create',
    }),
    confirm: telnyxMethod({
        method: 'POST',
        path: '/{porting_order_id}/actions/confirm',
        urlParams: ['porting_order_id'],
        paramsNames: ['porting_order_id'],
        methodType: 'create',
    }),
    share: telnyxMethod({
        method: 'POST',
        path: '/{porting_order_id}/actions/share',
        urlParams: ['porting_order_id'],
        paramsNames: ['porting_order_id'],
        methodType: 'create',
    }),
    createComment: telnyxMethod({
        method: 'POST',
        path: '/{porting_order_id}/comments',
        methodType: 'create',
        urlParams: ['porting_order_id'],
        paramsNames: ['porting_order_id'],
    }),
    listComments: telnyxMethod({
        method: 'GET',
        path: '/{id}/comments',
        methodType: 'list',
        urlParams: ['id'],
    }),
    listRequirements: telnyxMethod({
        method: 'GET',
        path: '/{id}/requirements',
        methodType: 'list',
        urlParams: ['id'],
    }),
    listVerificationCodes: telnyxMethod({
        method: 'GET',
        path: '/{id}/verification_codes',
        methodType: 'list',
        urlParams: ['id'],
    }),
    retrieveSubRequest: telnyxMethod({
        method: 'GET',
        path: '/{id}/sub_request',
        methodType: 'retrieve',
        urlParams: ['id'],
    }),
    sendVerificationCode: telnyxMethod({
        method: 'POST',
        path: '/{porting_order_id}/verification_codes/send',
        methodType: 'create',
        urlParams: ['porting_order_id'],
        paramsNames: ['porting_order_id'],
    }),
    verifyVerificationCode: telnyxMethod({
        method: 'POST',
        path: '/{porting_order_id}/verification_codes/verify',
        methodType: 'create',
        urlParams: ['porting_order_id'],
        paramsNames: ['porting_order_id'],
    }),
    listPhoneNumberBlocks: telnyxMethod({
        method: 'GET',
        path: '/{porting_order_id}/phone_number_blocks',
        methodType: 'list',
        urlParams: ['porting_order_id'],
    }),
    createPhoneNumberBlock: telnyxMethod({
        method: 'POST',
        path: '/{porting_order_id}/phone_number_blocks',
        methodType: 'create',
        urlParams: ['porting_order_id'],
        paramsNames: ['porting_order_id'],
    }),
    delPhoneNumberBlock: telnyxMethod({
        method: 'DELETE',
        path: '/{porting_order_id}/phone_number_blocks/{id}',
        urlParams: ['porting_order_id', 'id'],
        paramsNames: ['porting_order_id', 'id'],
    }),
    listPhoneNumberExtensions: telnyxMethod({
        method: 'GET',
        path: '/{porting_order_id}/phone_number_extensions',
        methodType: 'list',
        urlParams: ['porting_order_id'],
    }),
    createPhoneNumberExtension: telnyxMethod({
        method: 'POST',
        path: '/{porting_order_id}/phone_number_extensions',
        methodType: 'create',
        urlParams: ['porting_order_id'],
        paramsNames: ['porting_order_id'],
    }),
    delPhoneNumberExtension: telnyxMethod({
        method: 'DELETE',
        path: '/{porting_order_id}/phone_number_extensions/{id}',
        urlParams: ['porting_order_id', 'id'],
        paramsNames: ['porting_order_id', 'id'],
    }),
});
