// telnyx/dist/resources/PortoutRequests.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PortoutRequests = void 0;
const TelnyxResource_js_1 = __importDefault(require("../TelnyxResource.js"));
const telnyxMethod = TelnyxResource_js_1.default.method;
exports.PortoutRequests = TelnyxResource_js_1.default.extend({
    path: 'portouts',
    includeBasic: ['list', 'retrieve'],
    updateStatus: telnyxMethod({
        method: 'PATCH',
        path: '/{id}/{status}',
        urlParams: ['id', 'status'],
    }),
    listComments: telnyxMethod({
        method: 'GET',
        path: '/{id}/comments',
        urlParams: ['id'],
        methodType: 'list',
    }),
    createComment: telnyxMethod({
        method: 'POST',
        path: '/{id}/comments',
        urlParams: ['id'],
        methodType: 'create',
    }),
    retrieveRequest: telnyxMethod({
        method: 'GET',
        path: '/{id}',
        urlParams: ['id'],
        methodType: 'retrieve',
    }),
    retrieveComment: telnyxMethod({
        method: 'GET',
        path: '/{id}/comments',
        urlParams: ['id'],
        methodType: 'retrieve',
    }),
    listDocuments: telnyxMethod({
        method: 'GET',
        path: '/{id}/supporting_documents',
        urlParams: ['id'],
        methodType: 'list',
    }),
    createDocuments: telnyxMethod({
        method: 'POST',
        path: '/{id}/supporting_documents',
        urlParams: ['id'],
        methodType: 'create',
    }),
});
