// telnyx/dist/resources/Queues.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Queues = void 0;
const TelnyxResource_js_1 = __importDefault(require("../TelnyxResource.js"));
const utils = __importStar(require("../utils.js"));
const telnyxMethod = TelnyxResource_js_1.default.method;
const ACTIONS = ['calls'];
function getSpec(queueName) {
    return function (methodName) {
        return {
            method: 'GET',
            path: `/{queueName}/${methodName}`,
            urlParams: ['queueName'],
            paramsValues: [queueName],
            paramsNames: ['queueName'],
            methodType: 'list',
        };
    };
}
function transformResponseData(response, telnyx) {
    const methods = utils.createNestedMethods(telnyxMethod, ACTIONS, getSpec(response.data.name));
    methods.retrieveCall = telnyxMethod({
        method: 'GET',
        path: '/{queue_name}/calls/{call_control_id}',
        urlParams: ['queue_name', 'call_control_id'],
        paramsValues: [response.data.queue_name],
        paramsNames: ['queue_name', 'call_control_id'],
    });
    return utils.addResourceToResponseData(response, telnyx, 'queues', methods);
}
exports.Queues = TelnyxResource_js_1.default.extend({
    path: 'queues',
    retrieve: telnyxMethod({
        method: 'GET',
        path: '/{queue_name}',
        urlParams: ['queue_name'],
        transformResponseData: transformResponseData,
    }),
    listCalls: telnyxMethod({
        method: 'GET',
        path: '/{queue_name}/calls',
        urlParams: ['queue_name'],
    }),
    retrieveCall: telnyxMethod({
        method: 'GET',
        path: '/{queue_name}/calls/{call_control_id}',
        urlParams: ['queue_name', 'call_control_id'],
        paramsNames: ['queue_name', 'call_control_id'],
        methodType: 'retrieve',
    }),
});
