// telnyx/dist/resources/StorageBuckets.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageBuckets = void 0;
const TelnyxResource_js_1 = __importDefault(require("../TelnyxResource.js"));
const telnyxMethod = TelnyxResource_js_1.default.method;
exports.StorageBuckets = TelnyxResource_js_1.default.extend({
    path: 'storage/buckets',
    usage: telnyxMethod({
        method: 'GET',
        path: '/{bucketName}/usage/storage',
        urlParams: ['bucketName'],
    }),
    apiUsage: telnyxMethod({
        method: 'GET',
        path: '/{bucketName}/usage/api',
        urlParams: ['bucketName'],
    }),
    sslCertificate: telnyxMethod({
        method: 'GET',
        path: '/{bucketName}/ssl_certificate',
        urlParams: ['bucketName'],
    }),
    addSslCertificate: telnyxMethod({
        method: 'PUT',
        path: '/{bucketName}/ssl_certificate',
        urlParams: ['bucketName'],
    }),
    delSslCertificate: telnyxMethod({
        method: 'DELETE',
        path: '/{bucketName}/ssl_certificate',
        urlParams: ['bucketName'],
    }),
});
