// telnyx/dist/resources/Verifications.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Verifications = void 0;
const TelnyxResource_js_1 = __importDefault(require("../TelnyxResource.js"));
const utils = __importStar(require("../utils.js"));
const telnyxMethod = TelnyxResource_js_1.default.method;
const transformResponseData = (response, telnyx) => {
    return utils.addResourceToResponseData(response, telnyx, 'verifiedNumbers', {
        verify: telnyxMethod({
            method: 'POST',
            path: '/{verification_id}/actions/verify',
            urlParams: ['verification_id'],
            paramsNames: ['verification_id'],
            paramsValues: [response.data.id],
        }),
    });
};
exports.Verifications = TelnyxResource_js_1.default.extend({
    path: 'verifications',
    callVerify: telnyxMethod({
        method: 'POST',
        path: '/call',
        methodType: 'create',
    }),
    flashcallVerify: telnyxMethod({
        method: 'POST',
        path: '/flashcall',
        methodType: 'create',
    }),
    smsVerify: telnyxMethod({
        method: 'POST',
        path: '/sms',
        methodType: 'create',
    }),
    retrieve: telnyxMethod({
        method: 'GET',
        path: '/{verification_id}',
        urlParams: ['verification_id'],
        transformResponseData: transformResponseData,
    }),
    verify: telnyxMethod({
        method: 'POST',
        path: '/{verification_id}/actions/verify',
        urlParams: ['verification_id'],
        paramsNames: ['verification_id'],
    }),
    listByPhoneNumber: telnyxMethod({
        method: 'GET',
        path: '/by_phone_number/{phone_number}',
        urlParams: ['phone_number'],
        paramsNames: ['phone_number'],
        methodType: 'list',
    }),
    verifyByPhoneNumber: telnyxMethod({
        method: 'POST',
        path: '/by_phone_number/{phone_number}/actions/verify',
        urlParams: ['phone_number'],
        paramsNames: ['phone_number'],
        methodType: 'create',
    }),
});
