// telnyx/dist/resources/VirtualCrossConnects.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VirtualCrossConnects = void 0;
const TelnyxResource_js_1 = __importDefault(require("../TelnyxResource.js"));
const telnyxMethod = TelnyxResource_js_1.default.method;
exports.VirtualCrossConnects = TelnyxResource_js_1.default.extend({
    path: 'virtual_cross_connects',
    includeBasic: ['list', 'retrieve', 'delete', 'create'],
    ListVirtualCrossConnectRegions: telnyxMethod({
        method: 'GET',
        path: '/virtual_cross_connect/regions',
    }),
    DeleteVirtualCrossConnect: telnyxMethod({
        method: 'DELETE',
        path: '/virtual_cross_connects/{id}',
    }),
    GetVirtualCrossConnectRegion: telnyxMethod({
        method: 'GET',
        path: '/virtual_cross_connect/regions/{id}',
    }),
    ListVirtualCrossConnects: telnyxMethod({
        method: 'GET',
        path: '/virtual_cross_connects',
    }),
    ProvisionVirtualCrossConnect: telnyxMethod({
        method: 'POST',
        path: '/virtual_cross_connects/{id}/actions/provision',
    }),
});
