// telnyx/dist/resources/WireguardInterfaces.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WireguardInterfaces = void 0;
const TelnyxResource_js_1 = __importDefault(require("../TelnyxResource.js"));
const telnyxMethod = TelnyxResource_js_1.default.method;
exports.WireguardInterfaces = TelnyxResource_js_1.default.extend({
    path: 'wireguard_interfaces',
    includeBasic: ['list', 'retrieve', 'delete'],
    GetWireguardPeerAllowedIp: telnyxMethod({
        method: 'GET',
        path: '/wireguard_peers/{id}/allowed_ips/{child_id}',
    }),
    DeleteWireguardInterface: telnyxMethod({
        method: 'DELETE',
        path: '/wireguard_interfaces/{id}',
    }),
    DeleteWireguardPeer: telnyxMethod({
        method: 'DELETE',
        path: '/wireguard_peers/{id}',
    }),
    ListWireguardPeers: telnyxMethod({
        method: 'GET',
        path: '/wireguard_peers',
    }),
    GetWireguardPeerConfig: telnyxMethod({
        method: 'GET',
        path: '/wireguard_peers/{id}/config',
    }),
    ListWireguardInterfaces: telnyxMethod({
        method: 'GET',
        path: '/wireguard_interfaces',
    }),
    ListWireguardPeerAllowedIps: telnyxMethod({
        method: 'GET',
        path: '/wireguard_peers/{id}/allowed_ips',
    }),
});
